### PR TITLE
feat: editorial on Tree Painting

### PR DESCRIPTION
**Pull request by Yash-Singh1**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
image
did you now?

**Comment by Yash-Singh1** [comment]
> image
> 
> did you now?

Didn't see the general code conventions, fixed it now with more comments, description on $p_1$ and $p_2$ and tabs.

File: solutions/gold/cf-1187E.mdx [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+vector<vector<ll>> g;
+vector<ll> dp1, dp2;
+vector<ll> p1, p2;
+vector<ll> sz;
+
+// Calculate sz and dp1
+void dfs1(ll u, ll p) {
+	sz[u] = 1;
+	for (auto &v: g[u]) {
+		if (v == p) continue;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:93]: braces

File: solutions/gold/cf-1187E.mdx [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+vector<vector<ll>> g;
+vector<ll> dp1, dp2;
+vector<ll> p1, p2;
+vector<ll> sz;
+
+// Calculate sz and dp1
+void dfs1(ll u, ll p) {
+	sz[u] = 1;
+	for (auto &v: g[u]) {
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:92]: prefer if you didn't use auto

File: solutions/gold/cf-1187E.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/gold/cf-1187E.mdx:93]: in lotta other places asw

File: solutions/gold/cf-1187E.mdx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		if (dp2[v] + tot - dp1[v] > dp2[u]) {
+			dp2[u] = dp2[v] + tot - dp1[v];
+		}
+	}
+}
+
+int main() {
+	ios_base::sync_with_stdio(false);
+
+	ll n;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:133]: idt the nodes have to be `ll`?

File: solutions/gold/cf-1187E.mdx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		if (v == p) continue;
+		dfs3(v, u);
+		if (dp2[v] + tot - dp1[v] > dp2[u]) {
+			dp2[u] = dp2[v] + tot - dp1[v];
+		}
+	}
+}
+
+int main() {
+	ios_base::sync_with_stdio(false);
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:131]: remove if runs w/o

File: solutions/gold/cf-1187E.mdx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	dfs1(0, -1);
+
+	// Run second DFS to calculate p1 and p2
+	dfs2(0, -1);
+
+	// Run third DFS to calculate dp2
+	dfs3(0, -1);
+
+	// Find maximum dp1/dp2 among all nodes
+	ll ans{0};
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:155]:
```suggestion
	ll ans = 0;
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/cf-1187E.mdx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+vector<ll> sz;
+
+// Calculate sz and dp1
+void dfs1(int u, int p) {
+	sz[u] = 1;
+	for (int &v: g[u]) {
+		if (v == p) {
+			continue;
+      dfs1(v, u);
+		}
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:96]: sus

File: solutions/gold/cf-1187E.mdx [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+source: CF
+title: Tree Painting
+author: Yash Singh
+---
+
+[Official Editorial](https://codeforces.com/blog/entry/68111)
+
+## Explanation
+
+We notice that choosing a vertice splits the current connected component into
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:12]: > vertice

i think it's vertex

Comment (SansPapyrus683) [reply solutions/gold/cf-1187E.mdx:12]: also this first sentence,,, seems kinda obvious?

File: solutions/gold/cf-1187E.mdx [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+the childrens' subtrees and the rest of the tree excluding the current subtree.
+Another observation we can make is that choosing a leaf node is more optimal
+than choosing anything in the middle because that way we have the entire size of
+the rest of the tree as a part of the leftover connected component. This hints
+for a DP on Trees solution, we choose one leaf node, work our way up and work
+our way down for each of the siblings. For example, for the following tree:
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:18]: hb uh
> This motivates a tree DP solution where we [do stuff]

what does "work up"/"work down" mean
also siblings of what? the leaf node?

File: solutions/gold/cf-1187E.mdx [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+![Tree Visualization](cf-1187E/tree.png)
+
+If we select the 4th node as our starting point, then the process would look
+like the following:
+
+1. Work our way up to 3
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:25]: again, what does "work our way up" mean

File: solutions/gold/cf-1187E.mdx [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+Then we can calculate $p_1[u]$ and $p_2[u]$ using the following recurrence:
+
+$$
+p_1[u] = p_1[p] + sz[p] - sz[u] \newline
+p_2[u] = p_2[p] + p_1[u] + sz[p] - sz[u]
+$$
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(V+E)$
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:73]: put these in terms of the problem's constraints

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/cf-1187E.mdx (hunk 6 above)
Comment (Yash-Singh1) [reply solutions/gold/cf-1187E.mdx:96]: o lol i was editing in vscode it prob reset the spaces

File: solutions/gold/cf-1187E.mdx (hunk 8 above)
Comment (Yash-Singh1) [reply solutions/gold/cf-1187E.mdx:18]: ya i worded that really bad lemme see if i can make it better

File: solutions/gold/cf-1187E.mdx (hunk 9 above)
Comment (Yash-Singh1) [reply solutions/gold/cf-1187E.mdx:25]: Work our way up means to start from the leaf node and go through each of the ancestors.

Comment (SansPapyrus683) [reply solutions/gold/cf-1187E.mdx:25]: and...what are you doing when going through the ancestors?

File: solutions/gold/cf-1187E.mdx (hunk 8 above)
Comment (Yash-Singh1) [reply solutions/gold/cf-1187E.mdx]: I changed the wording a bit and the pic for the tree, could you check now?

File: solutions/gold/cf-1187E.mdx [hunk 11]
```diff
@@ -9,13 +9,13 @@ author: Yash Singh
...
 ## Explanation
 
-We notice that choosing a vertice splits the current connected component into
-the childrens' subtrees and the rest of the tree excluding the current subtree.
-Another observation we can make is that choosing a leaf node is more optimal
-than choosing anything in the middle because that way we have the entire size of
-the rest of the tree as a part of the leftover connected component. This hints
-for a DP on Trees solution, we choose one leaf node, work our way up and work
-our way down for each of the siblings. For example, for the following tree:
+An observation we can make is that choosing a leaf node is more optimal than
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1187E.mdx:12]:
```suggestion
Notice that choosing a leaf node is more optimal than
```

**Review by SansPapyrus683 — APPROVED** [review]
seems good enough

**Comment by SansPapyrus683** [comment]
@envyaims @ryanchou-dev review

File: solutions/gold/cf-1187E.mdx [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+		}
+		dfs2(v, u);
+	}
+}
+
+// Calculate dp2
+void dfs3(int u, int p) {
+	dp2[u] = (g[u].size() == 1 ? p2[u]+1 : 0);
+	ll tot = 0;
+	for (int &v: g[u]) {
```
Comment (envyaims) [inline solutions/gold/cf-1187E.mdx:124]: theres no reason to pass in ref for all these for loops

just a personal preference but it looks ugly lol

**Review by envyaims — APPROVED** [review]
ok

**Review by ryanchou-dev — APPROVED** [review]
address other stuff and then ok

**Comment by SansPapyrus683** [comment]
the references aren't too big of an issue anyways

**Pull request merged**